File: frontend/src/services/adminUsage.ts
import { authorizedFetch, parseFastApiDetail } from './auth';

/** accesses = logins + renovação de sessão (refresh). events_created = eventos de agenda criados. */
export interface UsageTotals {
  accesses: number;
  chat_completion: number;
  voice_session: number;
  events_created: number;
}

export interface UsageUserRow {
  user_id: string;
  name: string;
  email: string;
  accesses: number;
  chat_completion: number;
  voice_session: number;
  events_created: number;
}

export interface UsageSummary {
  period_from: string;
  period_to: string;
  totals: UsageTotals;
  users: UsageUserRow[];
}

export interface UsageTimeseriesRow {
  day: string;
  accesses: number;
  chat_completion: number;
  voice_session: number;
  events_created: number;
}

export interface UsageTimeseries {
  period_from: string;
  period_to: string;
  series: UsageTimeseriesRow[];
}

async function parseError(res: Response): Promise<string> {
  const body = await res.json().catch(() => ({}));
  return parseFastApiDetail(body, 'Erro na requisição');
}

function buildRangeQuery(dateFrom?: string, dateTo?: string): string {
  const params = new URLSearchParams();
  if (dateFrom) params.set('date_from', dateFrom);
  if (dateTo) params.set('date_to', dateTo);
  const q = params.toString();
  return q ? `?${q}` : '';
}

/** Omitir intervalo = últimos 30 dias (backend). */
export async function fetchUsageSummary(dateFrom?: string, dateTo?: string): Promise<UsageSummary> {
  const res = await authorizedFetch(`/admin/usage/summary${buildRangeQuery(dateFrom, dateTo)}`);
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function fetchUsageTimeseries(dateFrom?: string, dateTo?: string): Promise<UsageTimeseries> {
  const res = await authorizedFetch(`/admin/usage/timeseries${buildRangeQuery(dateFrom, dateTo)}`);
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

File: backend/presentation/api/schemas/admin_usage.py
from datetime import datetime

from pydantic import BaseModel, Field


class UsageTotals(BaseModel):
    """accesses = logins + renovações de sessão (token refresh). events_created = agenda (calendar_events)."""

    accesses: int = Field(ge=0)
    chat_completion: int = Field(ge=0)
    voice_session: int = Field(ge=0)
    events_created: int = Field(ge=0)


class UsageUserRow(BaseModel):
    user_id: str
    name: str
    email: str
    accesses: int = Field(ge=0)
    chat_completion: int = Field(ge=0)
    voice_session: int = Field(ge=0)
    events_created: int = Field(ge=0)


class UsageSummaryResponse(BaseModel):
    period_from: datetime
    period_to: datetime
    totals: UsageTotals
    users: list[UsageUserRow]


class UsageTimeseriesRow(BaseModel):
    day: str
    accesses: int = Field(ge=0)
    chat_completion: int = Field(ge=0)
    voice_session: int = Field(ge=0)
    events_created: int = Field(ge=0)


class UsageTimeseriesResponse(BaseModel):
    period_from: datetime
    period_to: datetime
    series: list[UsageTimeseriesRow]
